src/platform_impl/linux/x11/mod.rs

```diff
@@ -61,20 +62,79 @@ use self::{
     event_processor::EventProcessor,
     ime::{Ime, ImeCreationError, ImeReceiver, ImeRequest, ImeSender},
 };
-use super::common::xkb_state::KbdState;
+use super::{common::xkb_state::KbdState, OsError};
 use crate::{
-    error::OsError as RootOsError,
+    error::{OsError as RootOsError, RunLoopError},
     event::{Event, StartCause},
     event_loop::{ControlFlow, DeviceEvents, EventLoopClosed, EventLoopWindowTarget as RootELW},
+    platform::pump_events::PumpStatus,
     platform_impl::{
-        platform::{sticky_exit_callback, WindowId},
+        platform::{min_timeout, sticky_exit_callback, WindowId},
         PlatformSpecificWindowBuilderAttributes,
     },
     window::WindowAttributes,
 };
 
 type X11Source = Generic<RawFd>;
 
+struct WakeSender<T> {
+    sender: Sender<T>,
+    waker: Ping,
+}
+
+impl<T> Clone for WakeSender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sender: self.sender.clone(),
+            waker: self.waker.clone(),
+        }
+    }
+}
+
+impl<T> WakeSender<T> {
+    pub fn send(&self, t: T) -> Result<(), EventLoopClosed<T>> {
+        let res = self.sender.send(t).map_err(|e| EventLoopClosed(e.0));
+        if res.is_ok() {
+            self.waker.ping();
+        }
+        res
+    }
+}
```

Comment:
This looks like a reimplementation of the [`Sender`](https://docs.rs/calloop/latest/calloop/channel/struct.Sender.html) type from `calloop`. I don't see why the previous code that cached the events in `VecDeque`s was removed.

Reply:
A notable thing missing from the calloop channel is missing is the ability to peek the channel receiver.

I think there was a chicken/egg issue also with `has_pending` (which we want to know for short circuiting polling/dispatching) since it was based on state (the `VecDeque`s) filled in during dispatch.

Some of these changes were made while I was rebasing over the changes to add calloop to the x11 backend and I added some extra notes to the commit message:

```
    I found that the recent changes (4ac2006cbc5a) to port the X11 backend
    from mio to calloop lost the ability to check for pending events before
    needing to poll/dispatch. (The `has_pending` state being queried
    before dispatching() was based on state that was filled in during
    dispatching)

    As part of the rebase this re-introduces the PeekableReceiver and
    WakeSender which are small utilities on top of
    `std::sync::mpsc::channel()`. This adds a calloop `PingSource`
    so we can use a `Ping` as a generic event loop waker.

    For taking into account false positive wake ups the X11 source now
    tracks when the file descriptor is readable so after we poll via
    calloop we can then specifically check if there are new X11 events
    or pending redraw/user events when deciding whether to skip the
    event loop iteration.
```